tests/unit/s2n_psk_test.c

```diff
@@ -516,6 +516,136 @@ int main(int argc, char **argv)
 
             EXPECT_SUCCESS(s2n_connection_free(conn));
         }
+
+        /* Test s2n_connection_set_external_psks */
+        {
+            uint8_t identity_0[] = "identity";
+            uint8_t secret_0[] = "secret";
+
+            uint8_t identity_1[] = "identity 1";
+            uint8_t secret_1[] = "secret 1";
+
+            struct s2n_psk_identity first_identity = { identity_0, sizeof(identity_0) };
+            struct s2n_pre_shared_key first_psk = { first_identity, secret_0, sizeof(secret_0), S2N_PSK_HMAC_SHA384 };
+
+            struct s2n_psk_identity second_identity = { identity_1, sizeof(identity_1) };
+            struct s2n_pre_shared_key second_psk = { second_identity, secret_1, sizeof(secret_1), S2N_PSK_HMAC_SHA384 };
+
+            /* Safety checks */
+            {
+                EXPECT_FAILURE_WITH_ERRNO(s2n_connection_set_external_psks(NULL, &first_psk, 1), S2N_ERR_NULL);
+
+                struct s2n_connection *conn;
+                EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+                EXPECT_FAILURE_WITH_ERRNO(s2n_connection_set_external_psks(conn, NULL, 1), S2N_ERR_NULL);
+
+                EXPECT_FAILURE_WITH_ERRNO(s2n_connection_set_external_psks(conn, &first_psk, 0), S2N_ERR_SAFETY);
+
+                EXPECT_SUCCESS(s2n_connection_free(conn));
+            }
+
+            /* One psk */
+            {
+                struct s2n_connection *conn;
+                EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+                size_t num_psks = 1;
+
+                EXPECT_SUCCESS(s2n_connection_set_external_psks(conn, &first_psk, num_psks));
+                
+                struct s2n_psk *test_psk = NULL;
+                EXPECT_OK(s2n_array_get(&conn->psk_params.psk_list, 0, (void**) &test_psk));
+                EXPECT_NOT_NULL(test_psk);
+
+                EXPECT_EQUAL(test_psk->type, S2N_PSK_TYPE_EXTERNAL);
+                EXPECT_BYTEARRAY_EQUAL(test_psk->identity.data, first_identity.data, first_identity.length);
+                EXPECT_EQUAL(test_psk->identity.size, first_identity.length);
+                EXPECT_BYTEARRAY_EQUAL(test_psk->secret.data, first_psk.secret, first_psk.secret_len);
+                EXPECT_EQUAL(test_psk->secret.size, first_psk.secret_len);
```

Comment:
You should check the length before you check the contents of the array. Otherwise, what if first_psk.secret_len > test_psk->secret.size and you overflow while comparing?